File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque
    

def solution(n, computers):
    answer = 0
    
    networks = {c:set() for c in range(n)}
    
    for i in range(n):
        for j in range(n):
            if i == j:
                continue
            
            if computers[i][j] == 1:
                networks[i].add(j)
                networks[j].add(i)
            
    visited = [False] * n
            
    for c in networks.keys():
        if not visited[c]:
            # bfs
            answer += 1
            
            queue = deque([c])
            visited[c] = True
            
            while queue:
                nxt_c = queue.popleft()
                
                for adj_c in networks[nxt_c]:
                    if not visited[adj_c]:
                        queue.append(adj_c)
                        visited[adj_c] = True
    
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.kt
class Solution {
    fun solution(n: Int, computers: Array<IntArray>): Int {
        var answer = 0
        
        val network = mutableMapOf<Int, MutableList<Int>>()
        val visited = mutableSetOf<Int>()
        
        for (i in 0 until n) {
            for (j in 0 until n) {
                if (computers[i][j] != 1) continue
                
                network.getOrPut(i) {mutableListOf()}.add(j)
            }
        }
        
        fun dfs(curr_com: Int): Unit {
            visited.add(curr_com)
            
            for (adj_com in network.getValue(curr_com)) {
                if (adj_com !in visited) {
                    dfs(adj_com)
                }
            }
        }
        
        for (com in 0 until n) {
            if (com !in visited) {
                dfs(com)
                answer++
            }
        }
        
        return answer
    }
}